code/datums/observation/observation.dm

```diff
@@ -62,76 +62,241 @@
 	var/list/global_listeners  // Associative list of instances that listen to all events of this type (as opposed to events belonging to a specific source) and the proc to call.
 
 /decl/observ/New()
 	all_observable_events.events += src
-	event_sources = list()
-	..()
+	return ..()
 
-/decl/observ/proc/is_listening(var/event_source, var/datum/proc_owner, var/proc_call)
+/decl/observ/Del()
+	all_observable_events.events -= src
+	return ..()
+
+/decl/observ/proc/is_listening(var/event_source, var/datum/listener, var/proc_call)
+	// Return whether there are global listeners unless the event source is given.
+	if (!event_source)
+		return !!global_listeners
+
+	// Return whether anything is listening to a source, if no listener is given.
+	if (!listener)
+		return global_listeners || (event_sources && event_source in event_sources)
+
+	// Return false if there are no sources or nothing is associated with that source.
+	if (!(event_sources && event_source in event_sources))
+		return FALSE
+
+	// Get and check the listeners for the reuqested event.
 	var/listeners = event_sources[event_source]
-	if(!listeners)
+	if (!(listener in listeners))
 		return FALSE
 
-	var/stored_proc_call = listeners[proc_owner]
-	return stored_proc_call && (!proc_call || stored_proc_call == proc_call)
+	// Return true unless a specific callback needs checked.
+	if (!proc_call)
+		return TRUE
+
+	// Check if the specific callback exists.
+	var/list/callback = listeners[listener]
+	if (istype(callback))
+		return proc_call in callback
+	return proc_call == callback
 
 /decl/observ/proc/has_listeners(var/event_source)
-	var/list/listeners = event_sources[event_source]
-	return listeners && listeners.len
+	return is_listening(event_source)
 
-/decl/observ/proc/register(var/datum/event_source, var/datum/proc_owner, var/proc_call)
-	if(!(event_source && proc_owner && proc_call))
+/decl/observ/proc/register(var/datum/event_source, var/datum/listener, var/proc_call)
+	// Sanity checking.
+	if (!(event_source && listener && proc_call))
 		return FALSE
-	if(istype(event_source, /decl/observ))
+	if (istype(event_source, /decl/observ))
 		return FALSE
 
-	if(!istype(event_source, expected_type))
+	// Crash if the event source is the wrong type.
+	if (!istype(event_source, expected_type))
 		CRASH("Unexpected type. Expected [expected_type], was [event_source.type]")
 
-	var/listeners = event_sources[event_source]
-	if(!listeners)
+	// Perform some setup.
+	if (!event_sources)
+		event_sources = list()
+
+	var/list/listeners = event_sources[event_source]
+	if (!listeners)
 		listeners = list()
 		event_sources[event_source] = listeners
-	listeners[proc_owner] = proc_call
 
+	// If no callbacks are set or the callback is set to the value we want, return true..
+	var/list/callbacks = listeners[listener]
+	if (!callbacks || callbacks == proc_call)
+		listeners[listener] = proc_call
```

Comment:
Why not always have listeners[listener]/global_listeners[listener] be assigned a list of proc_calls, instead of having all these istype() checks and handling of the is/is-not case everywhere?